aspnetcore/performance/rate-limit.md

```diff
@@ -0,0 +1,186 @@
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_fixed":::
+
+The preceding code:
+* Calls [UseRateLimiter](/dotnet/api/microsoft.aspnetcore.builder.ratelimiterapplicationbuilderextensions.useratelimiter) to enable rate limiting.
+* Creates a fixed window limiter with a policy name of `"fixed"` and sets:
+* `permitLimit` to 4 and the time `window` to 12. A maximum of 4 requests per each 12-second window are allowed.
+* `queueProcessingOrder` to `QueueProcessingOrder.OldestFirst`.
+* `queueLimit` to 2.
+
+Apps should use [Configuration](xref:fundamentals/configuration/index) to set limiter options. The following code updates the preceding code using [`MyRateLimitOptions`](https://github.com/dotnet/AspNetCore.Docs.Samples/blob/main/fundamentals/middleware/rate-limit/WebRateLimitAuth/Models/MyRateLimitOptions.cs) for configuration:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_fixed2":::
+
+<a name="slide"></a>
+
+### Sliding window limiter
+
+A sliding window algorithm:
+
+* Is similar to the fixed window limiter but adds segments per window. The window slides one segment each segment interval. The segment interval is (window time)/(segments per window).
+* Limits the requests for a window to `permitLimit` requests.
+* Each time window is divided in `n` segments per window.
+* Requests taken from the expired time segment one window back (`n` segments prior to the current segment), are added to the current segment. We refer to the most expired time segment one window back as the expired segment.  Consider the following table which shows a sliding window limiter with a 30-second window, 3 segments per window and a limit of 100 requests:
+
+* The top row and first column shows the time segment.
+* The second row shows the remaining requests available. The remaining requests are available-requests+recycled.
+* The third and lower rows show the requests made at that time segment and recycled requests available from the prior expired segment.
+* From time 30 on, the request taken from the 3 times slots previous are added back to the request limit.
+
+| Time | 0  | 10  | 20 | 30 | 40 | 50 | 60 |
+| ---- | -- | --  | -- | -- | -- | -- | -- |
+| Available | 100-20+0=80 | 80-30+0=50  | 50-40+0=10 | 10-30+20=0 |0+30-10=20 | 20-10+40=50 | 50-35+30=45 |
+|  0    | -20            |                                  |  |  |  |  | |
+|  10   |               | -30                             |  |  |  |  | |
+|  20   |               |            | -40                    |  |  |  | |
+|  30   | **[+20]**     |            |         | -30        |  |  | |
+|  40   |          |**[+30]**|       |                 | -10   |  | |
+|  50   |          |           | **[+40]**  |            |               | -10  | |
+|  60   |          |           |            |  **[+30]**  |    |  | -35|
```

Comment:
I think this table is too confusing

Reply:
What about something like my updated table? If you don't like that approach, I'll delete it. I also added a couple bullets explaining the blue and red lines.

Reply:
Maybe slightly better, still not a fan though. Maybe let some other folks look at it and see if they understand?

I'm slightly biased, but I like how the algorithm was displayed [here](https://devblogs.microsoft.com/dotnet/announcing-rate-limiting-for-dotnet/#sliding-window-limit).

Reply:
> Maybe slightly better, still not a fan though. Maybe let some other folks look at it and see if they understand?
> 
> I'm slightly biased, but I like how the algorithm was displayed [here](https://devblogs.microsoft.com/dotnet/announcing-rate-limiting-for-dotnet/#sliding-window-limit).

I had a hard time following that, it struck me as a diagram written by someone very familiar with the algorithm.  At any rate the explanation and diagrams belong in https://github.com/dotnet/docs/issues/30426 which @IEvangelist owns, so he can own that.

IMAO, rate limiting is a gamechanger so I'm going to merge this so it can be live. I'll tweat it  with a challenge to come up with a better diagram.

Reply:
![2022-08-30 13 39 11](https://user-images.githubusercontent.com/3605364/187563723-bd99fd9b-7026-4142-9d4f-dda571a903b2.jpg)
@BrennanConroy  I tried the preceding prototype but it looked worse.